### PR TITLE
IsolatedMap extension method

### PR DESCRIPTION
**Pull request by MartinJohns**
Here's the method a bit cleaned up and documented. I'm not sure how I would go on about unit testing this, or what sample project would be worth showing. You got an idea?

### CODE REVIEW

**Comment by kevinchalet** [comment]
FYI, here's the `Isolate` extension I've used in my OpenIddict prototype:

``` csharp
public static IApplicationBuilder Isolate(
    [NotNull] this IApplicationBuilder app,
    [NotNull] Action<IApplicationBuilder> configuration,
    [NotNull] Action<IServiceCollection> serviceConfiguration) {
    var services = new ServiceCollection();

    // Retrieve the runtime services from the host provider.
    var manifest = CallContextServiceLocator.Locator.ServiceProvider.GetService<IRuntimeServices>();
    if (manifest != null) {
        foreach (var service in manifest.Services) {
            services.AddTransient(service, sp => CallContextServiceLocator.Locator.ServiceProvider.GetService(service));
        }
    }

    services.AddLogging();

    // Copy the services added by the hosting layer.
    services.AddInstance(app.ApplicationServices.GetRequiredService<IApplicationEnvironment>());
    services.AddInstance(app.ApplicationServices.GetRequiredService<IApplicationLifetime>());
    services.AddInstance(app.ApplicationServices.GetRequiredService<IHostingEnvironment>());
    services.AddInstance(app.ApplicationServices.GetRequiredService<ILoggerFactory>());
    services.AddInstance(app.ApplicationServices.GetRequiredService<IHttpContextAccessor>());
    services.AddInstance(app.ApplicationServices.GetRequiredService<TelemetrySource>());
    services.AddInstance(app.ApplicationServices.GetRequiredService<TelemetryListener>());

    serviceConfiguration(services);
    var provider = services.BuildServiceProvider();

    var builder = new ApplicationBuilder(null);
    builder.ApplicationServices = provider;

    builder.Use(next => async context => {
        var priorApplicationServices = context.ApplicationServices;
        var scopeFactory = provider.GetRequiredService<IServiceScopeFactory>();

        // Store the original request services in the current ASP.NET context.
        context.Items[typeof(IServiceProvider)] = context.RequestServices;

        try {
            using (var scope = scopeFactory.CreateScope()) {
                context.ApplicationServices = provider;
                context.RequestServices = scope.ServiceProvider;

                await next(context);
            }
        }
        finally {
            context.RequestServices = null;
            context.ApplicationServices = priorApplicationServices;
        }
    });

    configuration(builder);

    return app.Use(next => {
        // Run the rest of the pipeline in the original context,
        // with the services defined by the parent application builder.
        builder.Run(async context => {
            var priorApplicationServices = context.ApplicationServices;
            var scopeFactory = app.ApplicationServices.GetRequiredService<IServiceScopeFactory>();

            try {
                using (var scope = scopeFactory.CreateScope()) {
                    context.ApplicationServices = app.ApplicationServices;
                    context.RequestServices = scope.ServiceProvider;

                    await next(context);
                }
            }
            finally {
                context.RequestServices = null;
                context.ApplicationServices = priorApplicationServices;
            }
        });

        var branch = builder.Build();

        return context => branch(context);
    });
}
```

File: src/AspNet.Hosting.Extensions/ApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        ///   has the default services registered. It will not share the <see cref="ServiceCollection"/> from the
+        ///   originating app. The isolated map will be configured using the configuration methods of the
+        ///   <see cref="TStartup"/> class.
+        /// </summary>
+        /// <typeparam name="TStartup">
+        ///   The startup class used to configure the new app and the service collection.
+        /// </typeparam>
+        /// <param name="app">The application builder to create the isolated app from.</param>
+        /// <returns>The new pipeline with the isolated application integrated.</returns>
+        public static IApplicationBuilder Isolated<TStartup>(
```
Comment (kevinchalet) [inline src/AspNet.Hosting.Extensions/ApplicationBuilderExtensions.cs]: Have you considered naming it `Isolate`? Looks like the other extensions (Use, Map, Run) are all verbs.

**Comment by kevinchalet** [comment]
It looks good. Ping me when you're done with this PR :+1:

**Comment by MartinJohns** [comment]
I agree and renamed `Isolated` to `Isolate`. I left the comfort method `IsolatedMap` as it is. I fixed some more styling issues (but I still disagree with K&R ;-)).

**Comment by kevinchalet** [comment]
Merged: https://github.com/aspnet-contrib/AspNet.Hosting.Extensions/commit/07eddee9170b5be29ee1e6ac60e207653a1044e0

@MartinJohns danke! :tada:

**Comment by MartinJohns** [comment]
Yayyy. :tada:

**Pull request closed without merging**